### DISCUSSION TITLE
Configure --legacy-peer-deps with .npmrc

### DISCUSSION THREAD

**Discussion by kentcdodds** [npm/rfcs#283] [RRFC] [answered]
I'd love it if I could configure npm (globally and per project) with the `--legacy-peer-deps` functionality. This is especially helpful when using dependencies whose maintainers are no longer active but their projects support versions of peerDeps that are outside the version range specified.

Proposed config name: `legacyPeerDeps` and it could be a `Boolean`.

This could be in wherever `.npmrc` files typically go and npm would respect those values like it does with everything else.

**Comment by darcyclarke** [comment, marked as answer]
@kentcdodds Hey Kent! This should already be possible with setting the `legacy-peer-deps` config (ref. https://docs.npmjs.com/cli/v7/using-npm/config#legacy-peer-deps). You can define it as a boolean in your local project or global `.npmrc` file by running `npm config set legacy-peer-deps true` with an optional `-g` to set it globally.

**Reply by kentcdodds** [reply]
Oh, 🙃 whoops. Thank you!

**Reply by DesignByOnyx** [reply]
Just a note for anybody else who finds this:
- The `.npmrc` file must be created an updated manually. Running the `npm config ...` command will not create or update your project's `.npmrc` file.
- Using a `.npmrc` file is advantageous for the following:
    - you don't have to remember to type `--legacy-peer-deps=true` whenever you run `npm install`
    - it can be checked into source control (git) and it should work for your whole team, CI/CD, etc.
    - it won't affect other projects, whereas running the `npm config ...` command will affect all projects for a user.

**Reply by GeekyMonkey** [reply]
The .npmrc file should contain this line
    legacy-peer-deps=true

**Reply by toddiuszho** [reply]
> The `.npmrc` file must be created an updated manually. Running the `npm config ..`. command will not create or update your project's `.npmrc` file.

Depends on your version of **npm**. With **lts/gallum** (npm version 8.3.1), you can modify `CWD/.npmrc` with `npm config set --location=project legacy-peer-deps true`

**Reply by ljharb** [reply]
(note that lts/gallium describes node, being on a node version doesn't guarantee you're on any particular npm version)

**Reply by ryders** [reply]
Hi folks, sorry to be reviving an old thread. We use CI/CD to automatically deploy to feature environments and I just came across this after someone in our team installed a npm package locally with `--legacy-peer-deps`... as you can assume this change automatically failed deployment, and it seems this solution works.

**However**, I'm not sure that I'm ok with just accepting all "legacy / unmaintained" packages moving forward and I'm curious to hear your approach on how to best manage our app integrity and package versions.

Q1. Would it not make sense for --legacy-peer-deps to have some value (other than boolean) in order to formally allow _certain_ packages through, but not all?

Q2. is your preferred "best practice" to (a) always allow all updates (which will generally create some problems with packages not maintained regularly enough) or (b) lock all packages version and perform monthly or quarterly or yearly "upgrades"

@darcyclarke curious to here your thoughts mate.

Any reading material welcome.

**Reply by ljharb** [reply]
In general, I'd suggest not allowing any dependency that requires the flag - since that means your dep graph is invalid.

The only reliable way to keep your dep graph valid is to enforce that it remains so, by never using the legacy flag.

**Reply by ryders** [reply]
thank @ljharb -- am very much i agreement with that. It does mean however that any upgrades may yield packages that need to be replaced and/or wait until the maintainer also upgrades, right? The only way to avoid this would be for our team to 

- fix package+send pull request hoping for action
- fork + fix and use "our" version of the package
- use a different library an possibly refactor code

right?

**Reply by ljharb** [reply]
Yes, exactly right.

**Comment by hippo-guy** [comment]
I just ran the command `npm config set legacy-peer-deps=true` and it changed the global file instead. Is this expected?

**Reply by ljharb** [reply]
It should change `~/.npmrc`, and yes that's expected.

**Reply by ruyadorno** [reply]
there's a `location` option that can be used to specify changing the project `.npmrc` instead, e.g:
```
npm config --location=project set legacy-peer-deps=true
```

ref: https://docs.npmjs.com/cli/v9/commands/npm-config#location

**Reply by mindplay-dk** [reply]
for some reason, NPM 9 ignored the project-level setting - in fact, it appears to ignore anything in the first line of `.npmrc` for some reason... putting a `# comment` at the start of the file seems to resolve this.

**Comment by GirijaHegdeGit** [comment]
npm config set legacy-peer-deps true : what happens if I run this command

**Reply by entranced** [reply]
It will add `legacy-peer-deps=true` to your .npmrc file.
